### ISSUE TITLE
`Dataset.feature_name.__len__()` do not equal to `Dataset.num_feature()` after `add_features_from` another dataset

### ISSUE THREAD

**Issue by JiaRu2016** [microsoft/LightGBM#3221]
## How you are using LightGBM?

* Python package

## Environment info

Operating System: Ubuntu 18.04

CPU/GPU model: GPU

C++ compiler version:   gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

CMake version:  cmake version 3.17.1

Python version: 3.8

LightGBM version or commit hash:  on both master and tag=v2.3.1 can reproduce this issue

## Reproducible example(s)

```python
import lightgbm as lgb
import numpy as np

ds0 = lgb.Dataset(np.arange(12).reshape(3,4), feature_name=[f'X{i}' for i in range(4)])
ds0.construct()
len(ds0.feature_name), ds0.num_feature()   # 4, 4

ds1 = lgb.Dataset(np.arange(9).reshape(3,3), feature_name=[f'Z{i}' for i in range(3)])
ds1.construct()
len(ds1.feature_name), ds1.num_feature()   # 3, 3

ds0.add_features_from(ds1)    #  <---------------  may be a bug come from here

len(ds0.feature_name), ds0.num_feature()    # 4, 7

ds0.construct()           # even after `construct` this problem still exits
len(ds0.feature_name), ds0.num_feature()     # 4, 7
```

## Steps to reproduce

1. open a jupyter nobtbook
2. run codes above, should get the same result as me (if reproduceable)

**Comment by guolinke** [comment]
refer to https://github.com/microsoft/LightGBM/pull/2754

**Comment by JiaRu2016** [comment]
I have merged `fix-add-features` into current `master` branch, rebuild it and reinstall python package. However this problem still exits (exactly same output as before). Moreover when constructing dataset a Warning message raise:
```python
ds0 = lgb.Dataset(np.random.normal(size=(3,4)), label=[1,2,3], feature_name=[f'X{i}' for i in range(4)], free_raw_data=False)
ds0.construct()

[LightGBM] [Warning] There are no meaningful features, as all feature values are constant.
```

**Comment by guolinke** [comment]
@JiaRu2016 I think the feature name in python-package is not fixed yet in the PR, it needs to get from lightgbm.dll.
ping @StrikerRUS to confirm.

**Comment by guolinke** [comment]
@JiaRu2016 BTW, LightGBM will filter the feature, based on min_data (and min_data_per_bin). 
In your test case, the number of data is smaller than min_data and min_data_per_bin, you should set them both to 1 to run your test.

**Comment by JiaRu2016** [comment]
@guolinke you are right. Explicitly set `ds.feature_name = ds.get_feature_name()` will solve this problem.

**Comment by StrikerRUS** [comment]
Fixed via #2754.
